### PR TITLE
Fix fallback issues to handle inplace case

### PR DESCRIPTION
**Pull request by gujinghui**
Fix fallback issues to handle inplace case

### CODE REVIEW

**Comment by gujinghui** [comment]
@yinghai  pls help review

File: caffe2/ideep/operators/elementwise_sum_op.cc [hunk 1]
```diff
@@ -1,44 +1,73 @@
...
+        ideep::sum::compute(scales, inputs_itensor, *Y);
+      }
+      return true;
     }
 
-    return true;
+    if (InputSize() == 2)
+      return fallback_add_.Run(0);
+
+    return fallback_sum_.Run(0);
```
Comment (yinghai) [inline caffe2/ideep/operators/elementwise_sum_op.cc]: sum doesn't support 2-input case? If it does, we can just use `fallback_sum_` to simplify things.

Comment (gujinghui) [reply caffe2/ideep/operators/elementwise_sum_op.cc]: Not sure if sum supports broadcast add...

**Comment by gujinghui** [comment]
@yinghai 
Sorry, submitted wrong patch before.
The sum fallback should be reviewed in #15267.
So sorry for inconvenience.

File: caffe2/ideep/operators/operator_fallback_ideep.h [hunk 2]
```diff
@@ -99,12 +99,14 @@ class C10_EXPORT IDEEPFallbackOp final : public IDEEPOperator {
         }
       } else {
         VLOG(1) << "Input " << i << " is not ideep::tensor. Skipping copy.";
-        // Note(jiayq): This removes a const but conceptually
-        // local_input_blobs will only be used as const blob input for the
-        // base op so we are still fine.
-        local_input_blobs_[i]->ShareExternal(
-            const_cast<void *>(OperatorBase::Inputs()[i]->GetRaw()),
-            OperatorBase::Inputs()[i]->meta());
+        if (OperatorBase::Inputs()[i]->GetRaw() != local_input_blobs_[i]->GetRaw()) {
```
Comment (yinghai) [inline caffe2/ideep/operators/operator_fallback_ideep.h]: What if this condition is not met? What do we do?

Comment (gujinghui) [reply caffe2/ideep/operators/operator_fallback_ideep.h]: Do nothing. They are using same buffer, that is, already shared.

**Review by yinghai — CHANGES_REQUESTED** [review]
LGTM. Please rebase to resolve conflicts.

**Comment by gujinghui** [comment]
@yinghai 
rebased. pls help review.

**Comment by gujinghui** [comment]
@yinghai 
The failure is not caused by this PR.

> Jan 09 11:00:24 test_scatter_stress_cuda (__main__.ProcessGroupGlooTest) ... Process process 1:
Jan 09 11:00:24 Traceback (most recent call last):
Jan 09 11:00:24   File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
Jan 09 11:00:24     self.run()
Jan 09 11:00:24   File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
Jan 09 11:00:24     self._target(*self._args, **self._kwargs)
Jan 09 11:00:24   File "test_c10d.py", line 487, in _run
Jan 09 11:00:24     getattr(self, self.id().split(".")[2])()
Jan 09 11:00:24   File "test_c10d.py", line 453, in wrapper
Jan 09 11:00:24     fn(self)
Jan 09 11:00:24   File "test_c10d.py", line 50, in wrapper
Jan 09 11:00:24     return func(*args, **kwargs)
Jan 09 11:00:24   File "test_c10d.py", line 867, in test_scatter_stress_cuda
Jan 09 11:00:24     self._test_scatter_stress(inputs, lambda t: t.clone().cuda())
Jan 09 11:00:24   File "test_c10d.py", line 844, in _test_scatter_stress
Jan 09 11:00:24     work_handle.wait()
Jan 09 11:00:24 RuntimeError: [/var/lib/jenkins/workspace/third_party/gloo/gloo/transport/tcp/unbound_buffer.cc:72] Timed out waiting 1000ms for recv operation to complete
Jan 09 11:00:25 Process process 2:
Jan 09 11:00:25 Traceback (most recent call last):
Jan 09 11:00:25   File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
Jan 09 11:00:25     self.run()
Jan 09 11:00:25   File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
Jan 09 11:00:25     self._target(*self._args, **self._kwargs)
Jan 09 11:00:25   File "test_c10d.py", line 487, in _run
Jan 09 11:00:25     getattr(self, self.id().split(".")[2])()
Jan 09 11:00:25   File "test_c10d.py", line 453, in wrapper
Jan 09 11:00:25     fn(self)
Jan 09 11:00:25   File "test_c10d.py", line 50, in wrapper
Jan 09 11:00:25     return func(*args, **kwargs)
Jan 09 11:00:25   File "test_c10d.py", line 867, in test_scatter_stress_cuda
Jan 09 11:00:25     self._test_scatter_stress(inputs, lambda t: t.clone().cuda())
Jan 09 11:00:25   File "test_c10d.py", line 844, in _test_scatter_stress
Jan 09 11:00:25     work_handle.wait()
Jan 09 11:00:25 RuntimeError: [/var/lib/jenkins/workspace/third_party/gloo/gloo/transport/tcp/pair.cc:543] Connection closed by peer [127.0.0.1]:36555
Jan 09 11:00:25 Process process 3:
Jan 09 11:00:25 Traceback (most recent call last):
Jan 09 11:00:25   File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
Jan 09 11:00:25     self.run()
Jan 09 11:00:25   File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
Jan 09 11:00:25     self._target(*self._args, **self._kwargs)
Jan 09 11:00:25   File "test_c10d.py", line 487, in _run
Jan 09 11:00:25     getattr(self, self.id().split(".")[2])()
Jan 09 11:00:25   File "test_c10d.py", line 453, in wrapper
Jan 09 11:00:25     fn(self)
Jan 09 11:00:25   File "test_c10d.py", line 50, in wrapper
Jan 09 11:00:25     return func(*args, **kwargs)
Jan 09 11:00:25   File "test_c10d.py", line 867, in test_scatter_stress_cuda
Jan 09 11:00:25     self._test_scatter_stress(inputs, lambda t: t.clone().cuda())
Jan 09 11:00:25   File "test_c10d.py", line 844, in _test_scatter_stress
Jan 09 11:00:25     work_handle.wait()
Jan 09 11:00:25 RuntimeError: [/var/lib/jenkins/workspace/third_party/gloo/gloo/transport/tcp/unbound_buffer.cc:72] Timed out waiting 1000ms for recv operation to complete

**Review by yinghai — APPROVED** [review]
LGTM. Thanks!

**Pull request closed without merging**